compact_str/src/repr/intrinsics.rs

```diff
@@ -0,0 +1,108 @@
+use cfg_if::cfg_if;
+
+/// Checks if `discriminant` equals [`super::HEAP_MASK`], if so, moves `pointer_src` to
+/// `pointer_dst`, and `length_src` to `length_dst`.
+///
+/// # Special Intrinsics
+/// * `aarch64`, uses the `csel` instruction
+/// * `x86` and `x86_64`, uses the `cmovz`
+#[inline(always)]
+pub fn cmov_ptr_len(
+    discriminant: u8,
+    pointer_src: *const u8,
+    pointer_dst: &mut *const u8,
+    length_src: usize,
+    length_dst: &mut usize,
+) {
+    let discriminant = discriminant as usize;
+    cfg_if! {
+        if #[cfg(all(target_arch = "aarch64", not(miri)))] {
+            unsafe {
+                core::arch::asm! {
+                    "cmp {d}, 254",
+                    "csel {p}, {hp}, {sp}, EQ",
+                    "csel {l}, {hl}, {sl}, EQ",
+                    d = in(reg) discriminant,
+                    p = inlateout(reg) *pointer_dst,
+                    hp = in(reg) *pointer_dst,
+                    sp = in(reg) pointer_src,
+                    l = inlateout(reg) *length_dst,
+                    hl = in(reg) *length_dst,
+                    sl = in(reg) length_src,
+                    options(pure, nomem, nostack),
+                };
+            }
+        } else if #[cfg(all(any(target_arch = "x86", target_arch = "x86_64"), not(miri)))] {
+            unsafe {
+                core::arch::asm! {
+                    "cmp {d}, 254",
+                    "cmovnz {pd}, {ps}",
+                    "cmovnz {ld}, {ls}",
```

Comment:
cmov looks at the flag register to tell if it should do the mov. So the first mov should always be unconditional, i.e. a normal mov.

Reply:
Ah, I misinterpreted the code. You are conditionally overwriting a variable, not an register. OK, that's fine :)